milli/src/update/settings.rs

```diff
@@ -310,6 +325,67 @@ impl<'a, 't, 'u, 'i> Settings<'a, 't, 'u, 'i> {
         }
     }
 
+    fn update_synonyms(&mut self) -> anyhow::Result<bool> {
+        match self.synonyms {
+            Setting::Set(ref synonyms) => {
+                fn normalize(analyzer: &Analyzer<&[u8]>, text: &str) -> Vec<String> {
+                    analyzer
+                        .analyze(text)
+                        .tokens()
+                        .filter_map(|token|
+                            if token.is_word() { Some(token.text().to_string()) } else { None }
+                        )
+                        .collect::<Vec<_>>()
+                }
+
+                let old_synonyms = self.index.synonyms(self.wtxn)?;
+
+                let mut config = AnalyzerConfig::default();
+
+                let stop_words = self.index.stop_words(self.wtxn)?;
+                if let Some(stop_words) = &stop_words {
+                    config.stop_words(stop_words);
+                }
+
+                let analyzer = Analyzer::new(config);
+
+                let new_synonyms = synonyms
+                    .into_iter()
+                    .map(|(word, synonyms)| {
+                        let normalized_word = normalize(&analyzer, word);
+                        let normalized_synonyms = synonyms
+                            .iter()
+                            .map(|synonym| normalize(&analyzer, synonym))
+                            .collect::<Vec<_>>();
+
+                        (normalized_word, normalized_synonyms)
+                    })
+                    .group_by(|(word, _)| word.clone())
+                    .into_iter()
+                    .map(|(word, group)| {
+                        let synonyms = group
+                            .map(|(_, synonyms)| synonyms)
+                            .concat()
+                            .into_iter()
+                            .unique()
+                            .collect();
+
+                        (word, synonyms)
+                    })
+                    .collect();
```

Comment:
```suggestion
                let mut new_synonyms = HashMap::new();
                for (word, synonyms) in synonyms {
                    // Normalize both the word and associated synonyms.
                    let normalized_word = normalize(&analyzer, word);
                    let normalized_synonyms = synonyms.iter().map(|synonym| normalize(&analyzer, synonym));

                    // Store the normalized synonyms under the normalized word,
                    // merging the possible duplicate words.
                    let synonyms = new_synonyms.entry(&normalized_word).or_default();
                    synonyms.extend(normalized_synonyms);

                    // Make sure that we don't have duplicate synonyms.
                    synonyms.sort_unstable();
                    synonyms.dedup();
                }
```

Could you please check that it is correct? I think it is easier to understand the meaning of this for loop than the multiple combinators method :)

Reply:
seems expensive to sort and dedup on every iteration. How about this?
```rust
let mut new_synonyms: HashMap<_, Vec<Vec<String>>> = HashMap::new();
for (word, synonyms) in synonyms {
    // Normalize both the word and associated synonyms.
    let normalized_word = normalize(&analyzer, word);
    let normalized_synonyms = synonyms
        .iter()
        .map(|synonym| normalize(&analyzer, synonym));

    // Store the normalized synonyms under the normalized word,
    // merging the possible duplicate words.
    let entry = new_synonyms.entry(normalized_word).or_default();
    entry.extend(normalized_synonyms);
}

// Make sure that we don't have duplicate synonyms.
new_synonyms = new_synonyms
    .into_iter()
    .map(|(word, synonyms)|
        (word, synonyms.into_iter().unique().collect())
    )
    .collect();
```

Also, implicit type annotation for `V` type of `new_synonyms`'s `HashMap` is needed
```
error[E0282]: type annotations needed for `HashMap<Vec<std::string::String>, V>`
   --> milli/src/update/settings.rs:358:69
    |
348 |                 let mut new_synonyms = HashMap::new();
    |                     ---------------- consider giving `new_synonyms` the explicit type `HashMap<Vec<std::string::String>, V>`, where the type parameter `V` is specified
...
358 |                     let entry = new_synonyms.entry(normalized_word).or_default();
    |                                                                     ^^^^^^^^^^ cannot infer type for type parameter `V`
    |
    = note: type must be known at this point
```

Reply:
I would prefer that we don't create a new HashMap for the sole purpose of depuplicating the values, maybe a simple `new_synonyms.iter_mut().for_each().sort_unstable().dedup()` after the loop could do the job?

Reply:
sounds good. Fixed.